### DISCUSSION TITLE
Is there a workaround to have a union of types in the discriminate union key field?

### DISCUSSION THREAD

**Discussion by HynekS** [colinhacks/zod#2018] [Q&A] [answered]
Is there a workaround to have a union of types in the discriminate union key field? I am using Zod for the first time in real life, and it feels great, but I've run into an issue. I want to create a discriminated union for "published posts" and "drafts". "Drafts" would have the "draft: true" field, while "published posts" would not (and their schema will be much more strict).

The problem is that the discriminate key field accepts only singular literal values, so I can't use a union type. **Is there another way to construct a discriminate union from a variant with a specific "key: value" and a collection of all other variants**?

```ts
const publishedPostSchema = z.object({
  title: z.string(),
  author: z.enum(["John Doe", "Jane Doe"]),
  dateCreated: z.string().transform((str) => new Date(str)),
  description: z.string(),
  draft: z.literal(null), // I would like to allow "falsy" values in general: `null`, `undefined`, `false`
});

const draftPostSchema = z.object({
  title: z.string().nullable(),
  author: z.enum(["John Doe", "Jane Doe"]).nullable(),
  dateCreated: z
    .string()
    .transform((str) => new Date(str))
    .nullable(),
  draft: z.literal(true),
});

const exampleBlogCollection = defineCollection({
  schema: z.discriminatedUnion("draft", [publishedPostSchema, draftPostSchema]),
});
```

**Comment by JacobWeisenburger** [comment, marked as answer]
I don't think what you are asking for is possible. Also, in my opinion, it defeats the purpose of using a discriminated union if you have more than 1 value for each object of the union.

**Reply by HynekS** [reply]
Thank you for replying. Yes, having more than one value is probably not optimal. For may specific use case, the solution would be rather simple: to allow coercing or refining the value *before* it will be used as a discriminant.

I've created a workaround by creating a base "published post" schema and extending it with all corresponding variants (`false`, `null`, `undefined`). Then I used a type predicate (`is`) to determine the proper union member and let TS do the inferrence correctly.
